packages/js-sdk/src/user/loginWithMFA.ts

```diff
@@ -0,0 +1,168 @@
+  });
+
+  const response = await request.execute();
+  response.data.user._kmd.authtoken = response.data.authToken;
+  return response.data;
+}
+
+async function completeMFALoginRetryable(
+  mfaComplete: (authenticator: string, context: MFAContext) => Promise<MFACompleteResult>,
+  context: MFAContext
+): Promise<any> {
+  const errMsgNoCode = 'MFA code is missing.';
+  const errMsgTrustDevice = 'trustDevice should be boolean.';
+  try {
+    const mfaCompleteResult = await mfaComplete(context.authenticator.id, context);
+    if (!isObjectLike(mfaCompleteResult) || mfaCompleteResult.code == null) {
+      throw new KinveyError(errMsgNoCode);
+    }
+
+    if (mfaCompleteResult.trustDevice != null && !isBoolean(mfaCompleteResult.trustDevice)) {
+      throw new KinveyError(errMsgTrustDevice);
+    }
+    const trustDevice = mfaCompleteResult.trustDevice || false;
+    const mfaData = await executeCompleteRequest(mfaCompleteResult.code, trustDevice);
+    return mfaData;
+  } catch (err) {
+    if (err.message === errMsgNoCode || err.message === errMsgTrustDevice || !(err instanceof KinveyError)) {
+      throw err;
+    }
+
+    context.retries += 1; // eslint-disable-line no-param-reassign
+    context.error = err; // eslint-disable-line no-param-reassign
+    return completeMFALoginRetryable(mfaComplete, context);
+  }
+}
+
+export interface MFACompleteResult {
+  code: string;
+  trustDevice?: boolean;
+}
+
+async function _loginWithMFA(
+  username: string,
+  password: string,
+  selectAuthenticator: (authenticators: object[], context: MFAContext) => Promise<string>,
+  mfaComplete: (authenticator: string, context: MFAContext) => Promise<MFACompleteResult>,
+  options: LoginOptions = {}
+): Promise<User> {
+  validateNoActiveUser();
+
+  const credentials = validateCredentials(username, password);
+  if (!selectAuthenticator) {
+    throw new KinveyError('Function to select authenticator is missing.');
+  }
+
+  if (!mfaComplete) {
+    throw new KinveyError('Function to complete MFA is missing.');
+  }
+
+  const userHasDeviceToken = hasDeviceToken(credentials.username);
+  if (userHasDeviceToken) {
+    credentials.deviceToken = getDeviceToken(credentials.username);
+  }
+
+  const loginResult = await executeLoginRequest(credentials, options.timeout);
+  if (!loginResult.mfaRequired) {
+    setSession(loginResult.user);
+    return new User(loginResult.user);
+  }
+
+  if (userHasDeviceToken) {
+    // MFA is still required which means that the device token has expired
+    removeDeviceToken(credentials.username);
+  }
+
+  setMFASessionToken(loginResult.mfaSessionToken);
+  if (loginResult.authenticators.length === 0) {
+    throw new KinveyError(errMsgNoAuthenticators);
+  }
+
+  const context = {
+    mfaSessionToken: loginResult.mfaSessionToken,
+    retries: 0,
+    authenticator: null,
+  };
+  const selectedAuthenticatorId = await selectAuthenticator(loginResult.authenticators, context);
+  if (selectedAuthenticatorId == null) {
```

Comment:
Any non null result from the select will bypass the check. This check should be done after we find the authenticator from the authenticators list.

Reply:
We care about the null/undefined scenario because this is the way to stop the login process. If the end user decides to abandon the login while selecting the authenticator, then the developer should return null.
If selectAuthenticator returns a non-existent ID, then the backend will return a NotFound error which I think will be a rare scenario. Maybe, it could happen during development, so I don't think we should handle it client-side.

Reply:
but we already know the authenticators from the loginResult here:
`context.authenticator = loginResult.authenticators.find((a) => a.id === selectedAuthenticatorId)`

my suggestion is simply to move the check after the above code and make it `if (!context.authenticator) { ... }` as we don't need to go to the backend in this scenario.